File: web_app/lib/market_time.ts
/**
 * Market-local date reckoning — the client-side twin of `src/server/calendar_events.py`.
 *
 * Every calendar date Investa shows belongs to a market, not to the viewer: an
 * earnings report is "today" while it is still today on that exchange, whether
 * the browser is in Bangkok or Los Angeles. Counting days against the device
 * clock instead is what made a US report a day out for a UTC+7 viewer.
 */

/** Fallback zone when an event does not name its exchange's (Investa is US-first). */
export const DEFAULT_MARKET_TIMEZONE = 'America/New_York';

/** `YYYY-MM-DD` for an ISO date/datetime string, or null if unparseable. */
function calendarDay(iso: string | null | undefined): string | null {
    if (typeof iso !== 'string') return null;
    const day = iso.slice(0, 10);
    return /^\d{4}-\d{2}-\d{2}$/.test(day) ? day : null;
}

/** Today's `YYYY-MM-DD` in a market's own timezone. */
export function marketToday(timeZone?: string | null): string {
    const now = new Date();
    for (const tz of [timeZone, DEFAULT_MARKET_TIMEZONE]) {
        if (!tz) continue;
        try {
            const parts = new Intl.DateTimeFormat('en-US', {
                timeZone: tz,
                year: 'numeric',
                month: '2-digit',
                day: '2-digit',
            }).formatToParts(now);
            const at = (type: string) => parts.find(p => p.type === type)?.value ?? '';
            const day = `${at('year')}-${at('month')}-${at('day')}`;
            if (calendarDay(day)) return day;
        } catch {
            // Unknown zone (or an engine without full ICU) — fall through to the default.
        }
    }
    // Last resort: the device's own date, which is what we were trying to avoid.
    return now.toISOString().slice(0, 10);
}

/**
 * Whole days from today-in-`timeZone` to a calendar date. Negative for the past,
 * null if the date can't be read.
 *
 * Both sides are pinned to UTC midnight so the subtraction can't be skewed by a
 * DST transition falling between them.
 */
export function marketDayDiff(iso: string | null | undefined, timeZone?: string | null): number | null {
    const day = calendarDay(iso);
    if (!day) return null;
    const target = Date.parse(`${day}T00:00:00Z`);
    const today = Date.parse(`${marketToday(timeZone)}T00:00:00Z`);
    if (isNaN(target) || isNaN(today)) return null;
    return Math.round((target - today) / 86400000);
}

/**
 * Whether a calendar date falls no later than `months` months past today in a
 * market's own timezone — the horizon filter behind the "3 months / 1 year"
 * calendar toggles.
 *
 * A month here is a calendar month, so the cutoff lands on the same day-of-month
 * (overflowing as JS does: Nov 30 + 3 months reaches into early March).
 */
export function isWithinMarketMonths(
    iso: string | null | undefined,
    months: number,
    timeZone?: string | null,
): boolean {
    const day = calendarDay(iso);
    if (!day) return false;
    const target = Date.parse(`${day}T00:00:00Z`);
    if (isNaN(target)) return false;
    const [year, month, date] = marketToday(timeZone).split('-').map(Number);
    return target <= Date.UTC(year, month - 1 + months, date);
}

/**
 * Format a date-only string ("2026-08-05") for display.
 *
 * Rendered in UTC on purpose: the value is a calendar day, not an instant, so
 * letting the browser localize it would slide it to the previous day for anyone
 * west of UTC.
 *
 * Pinned to the Gregorian calendar for the same reason the zone is pinned: the
 * day belongs to the market, not to the reader's device. A browser set to Thai
 * defaults to the Buddhist era and renders a fiscal quarter ending March 2026
 * as "2569". Month names stay localized.
 */
export function formatCalendarDate(
    iso: string | null | undefined,
    options?: Intl.DateTimeFormatOptions,
): string {
    const day = calendarDay(iso);
    if (!day) return typeof iso === 'string' ? iso : '';
    const d = new Date(`${day}T00:00:00Z`);
    if (isNaN(d.getTime())) return day;
    if (options) {
        return d.toLocaleDateString(undefined, { calendar: 'gregory', ...options, timeZone: 'UTC' });
    }
    // `DD MMM YYYY`, assembled rather than delegated to `toLocaleDateString`,
    // which orders the parts by the viewer's locale: the same lot would read
    // "Aug 5, 2026" for one reader and "05/08/2026" for another, and neither
    // would match the native clients. Month names stay localized; only the
    // order and the zero-padded day are fixed, so dates line up down a column.
    return `${pad2(d.getUTCDate())} ${shortMonth(d)} ${d.getUTCFullYear()}`;
}

/** `DD MMM` — the same notation with the year dropped, for an axis or a row
 *  with no room for it. Day-first like the full form; never "Aug 5". */
export function formatCalendarDayMonth(iso: string | null | undefined): string {
    const day = calendarDay(iso);
    if (!day) return typeof iso === 'string' ? iso : '';
    const d = new Date(`${day}T00:00:00Z`);
    if (isNaN(d.getTime())) return day;
    return `${pad2(d.getUTCDate())} ${shortMonth(d)}`;
}

function pad2(n: number): string {
    return String(n).padStart(2, '0');
}

function shortMonth(d: Date): string {
    return d.toLocaleDateString(undefined, { calendar: 'gregory', month: 'short', timeZone: 'UTC' });
}

File: macos_app/Investa/App/MarketTime.swift
import Foundation

/// Market-local date reckoning — the SwiftUI twin of `web_app/lib/market_time.ts`
/// and `src/server/calendar_events.py`.
///
/// Every calendar date Investa shows belongs to a market, not to the device: an
/// earnings report is "today" while it is still today on that exchange, whether
/// the phone is in Bangkok or Los Angeles. Counting days against `Calendar.current`
/// instead is what put a US report a day out for a viewer in Asia.
enum MarketTime {
    /// Fallback zone for events that do not name their exchange's (Investa is US-first).
    static let defaultTimeZoneIdentifier = "America/New_York"

    private static let utc = TimeZone(identifier: "UTC") ?? TimeZone(secondsFromGMT: 0)!

    /// Calendar days are compared in UTC so that no exchange's DST transition can
    /// land between the two dates being subtracted.
    private static let utcCalendar: Calendar = {
        var cal = Calendar(identifier: .gregorian)
        cal.timeZone = utc
        return cal
    }()

    private static let dayParser: DateFormatter = {
        let f = DateFormatter()
        f.locale = Locale(identifier: "en_US_POSIX")
        f.dateFormat = "yyyy-MM-dd"
        f.timeZone = utc
        return f
    }()

    /// A market date belongs to the market's calendar, not the device's. A
    /// phone set to Thailand defaults to the Buddhist era and renders a fiscal
    /// quarter ending March 2026 as "Mar 2569 BE" — the same class of mistake
    /// as reading a US market date on a Bangkok clock, and just as wrong on a
    /// filed statement. Pinning the calendar keeps the year the filer's own
    /// while leaving month names localized.
    private static let gregorian = Calendar(identifier: .gregorian)

    /// The reader's own locale with the Gregorian calendar forced on.
    ///
    /// A device set to a Thai region defaults to the Buddhist era, so a
    /// formatter that inherits it renders 23 August 2018 as "23 Aug 2561" — a
    /// year no market, filing or trade confirmation has ever used. Setting only
    /// `DateFormatter.calendar` is not enough on its own: a formatter takes its
    /// calendar from its locale, so the calendar has to be baked into the
    /// locale. Month and weekday names stay in the reader's language; only the
    /// era changes.
    static let displayLocale: Locale = gregorianLocale(.current)

    /// `base` with the Gregorian calendar substituted, and nothing else touched.
    static func gregorianLocale(_ base: Locale) -> Locale {
        if base.calendar.identifier == .gregorian { return base }
        var components = Locale.Components(locale: base)
        components.calendar = .gregorian
        return Locale(components: components)
    }

    /// The Gregorian calendar on the device's own zone — for arithmetic on dates
    /// the reader picked ("a year ago", "the start of this year").
    /// `Calendar.current` is the calendar the device *displays*, so on a Thai
    /// device `component(.year:)` answers 2569 and any comparison against a year
    /// parsed from the API is off by 543.
    static var localCalendar: Calendar {
        var cal = gregorian
        cal.timeZone = .current
        return cal
    }

    /// The US market's zone — the default for an axis or tooltip plotting a
    /// series that does not name its own exchange.
    static var defaultZone: TimeZone { zone(nil) }

    /// A display formatter pinned to the Gregorian calendar, and to a market's
    /// zone when one is given. Every user-visible date the app builds outside
    /// this file goes through here: a bare `DateFormatter()` inherits the
    /// device's calendar and prints the wrong year.
    static func formatter(_ pattern: String, timeZone: TimeZone? = nil) -> DateFormatter {
        let f = DateFormatter()
        // Locale first — assigning one resets `calendar` to the locale's own.
        f.locale = displayLocale
        f.calendar = gregorian
        f.dateFormat = pattern
        if let timeZone { f.timeZone = timeZone }
        return f
    }

    /// A formatter for the ISO `yyyy-MM-dd` the API speaks — fixed English and
    /// Gregorian, since this is a wire format, not a notation a reader sees.
    /// A device calendar leaking in here would send the backend "2569-08-25".
    static func isoFormatter(timeZone: TimeZone? = nil) -> DateFormatter {
        let f = DateFormatter()
        f.locale = Locale(identifier: "en_US_POSIX")
        f.calendar = gregorian
        f.dateFormat = "yyyy-MM-dd"
        if let timeZone { f.timeZone = timeZone }
        return f
    }

    /// "29 Jul 2026" — `DD MMM YYYY`, the app's one date notation.
    ///
    /// An explicit pattern rather than `dateStyle = .medium`, which orders the
    /// parts by the device's locale: the same lot would read "Jul 29, 2026" on
    /// a US phone and "29/07/2026" on a British one, and a list of dates that
    /// changes shape with the reader is not a format. Month names stay
    /// localized; only the order and the zero-padded day are fixed, so figures
    /// line up down a column.
    ///
    /// Renders in UTC to match the parser — a calendar day is a day, not an
    /// instant, and must not slide for a viewer west of UTC.
    private static let mediumFormatter: DateFormatter = {
        let f = DateFormatter()
        f.locale = displayLocale
        f.calendar = gregorian
        f.dateFormat = "dd MMM yyyy"
        f.timeZone = utc
        return f
    }()

    /// "29 Jul" — the same notation with the year dropped, for a row or axis
    /// with no room for it. Day-first like the full form; never "Jul 29".
    private static let shortDayFormatter: DateFormatter = {
        let f = DateFormatter()
        f.locale = displayLocale
        f.calendar = gregorian
        f.dateFormat = "dd MMM"
        f.timeZone = utc
        return f
    }()

    /// "Mar 2026" — how a quarter is named, since four columns a year would
    /// otherwise all read "2026".
    ///
    /// Fixed English rather than localized, matching `yearFormatter` below: a
    /// localized month-year template carries an era marker in some regions
    /// ("Mar 2026 AD" for en_TH), and an axis label has no room for it.
    private static let monthYearFormatter: DateFormatter = {
        let f = DateFormatter()
        f.locale = Locale(identifier: "en_US_POSIX")
        f.calendar = gregorian
        f.dateFormat = "MMM yyyy"
        f.timeZone = utc
        return f
    }()

    /// "2026" — the axis label for a chart plotted on calendar days.
    private static let yearFormatter: DateFormatter = {
        let f = DateFormatter()
        f.locale = Locale(identifier: "en_US_POSIX")
        f.calendar = gregorian
        f.dateFormat = "yyyy"
        f.timeZone = utc
        return f
    }()

    /// The calendar day an ISO date (or datetime) string names, at UTC midnight.
    static func calendarDay(_ iso: String) -> Date? {
        dayParser.date(from: String(iso.prefix(10)))
    }

    /// The exchange's zone, falling back to the US default for a missing or
    /// unrecognized identifier.
    static func zone(_ identifier: String?) -> TimeZone {
        if let trimmed = identifier?.trimmingCharacters(in: .whitespaces), !trimmed.isEmpty,
           let tz = TimeZone(identifier: trimmed) {
            return tz
        }
        return TimeZone(identifier: defaultTimeZoneIdentifier) ?? utc
    }

    /// Today's calendar day on a market's own clock, at UTC midnight so it can be
    /// differenced against `calendarDay(_:)`.
    static func today(timeZone identifier: String?) -> Date? {
        var marketCalendar = Calendar(identifier: .gregorian)
        marketCalendar.timeZone = zone(identifier)
        let parts = marketCalendar.dateComponents([.year, .month, .day], from: Date())
        return utcCalendar.date(from: DateComponents(year: parts.year, month: parts.month, day: parts.day))
    }

    /// Whole days from today-on-the-market to `iso`. Negative for the past, nil if
    /// the date can't be read.
    static func dayDiff(_ iso: String, timeZone identifier: String?) -> Int? {
        guard let target = calendarDay(iso), let today = today(timeZone: identifier) else { return nil }
        return utcCalendar.dateComponents([.day], from: today, to: target).day
    }

    /// Whether a calendar date falls no later than `months` months past today on a
    /// market's own clock — the horizon behind the "3 Months / 1 Year" calendar
    /// toggles. A month is a calendar month, clamped to the month's last day.
    static func isWithin(_ iso: String, months: Int, timeZone identifier: String?) -> Bool {
        guard let target = calendarDay(iso), let today = today(timeZone: identifier),
              let cutoff = utcCalendar.date(byAdding: .month, value: months, to: today)
        else { return false }
        return target <= cutoff
    }

    /// "29 Jul 2026" for a date-only string; the input is returned unchanged if
    /// it isn't a date.
    static func formatted(_ iso: String) -> String {
        guard let d = calendarDay(iso) else { return iso }
        return mediumFormatter.string(from: d)
    }

    /// "29 Jul 2026" for a day already parsed by `calendarDay(_:)`. Only for
    /// calendar days: an intraday timestamp would render in UTC and can name
    /// the wrong day.
    static func formatted(_ day: Date) -> String {
        mediumFormatter.string(from: day)
    }

    /// "29 Jul" for a date-only string.
    static func shortDay(_ iso: String) -> String {
        guard let d = calendarDay(iso) else { return iso }
        return shortDayFormatter.string(from: d)
    }

    /// "29 Jul" for a day already parsed by `calendarDay(_:)`.
    static func shortDay(_ day: Date) -> String {
        shortDayFormatter.string(from: day)
    }

    /// "Jul 2026" for a date-only string.
    static func monthYear(_ iso: String) -> String {
        guard let d = calendarDay(iso) else { return iso }
        return monthYearFormatter.string(from: d)
    }

    /// "Jul 2026" for a day already parsed by `calendarDay(_:)` — the form a
    /// chart axis hands back.
    static func monthYear(_ day: Date) -> String {
        monthYearFormatter.string(from: day)
    }

    /// The calendar year of a day produced by `calendarDay(_:)`, read back in the
    /// same zone it was built in.
    static func year(_ day: Date) -> String {
        yearFormatter.string(from: day)
    }
}
